lib/src/block_parser.dart

```diff
@@ -728,7 +730,137 @@ class ParagraphSyntax extends BlockSyntax {
       parser.advance();
     }
 
-    var contents = parser.document.parseInline(childLines.join('\n'));
-    return new Element('p', contents);
+    var paragraphLines = _extractReflinkDefinitions(parser, childLines);
+    if (paragraphLines == null) {
+      // Paragraph consisted solely of reference link definitions.
+      return new Text('');
+    } else {
+      var contents = new UnparsedContent(paragraphLines.join('\n'));
+      return new Element('p', [contents]);
+    }
+  }
+
+  // Extract reference link definitions from the front of the paragraph, and
+  // return the remaining paragraph lines.
+  List<String> _extractReflinkDefinitions(
+      BlockParser parser, List<String> lines) {
```

Comment:
This function is pretty hairy. (Fun fact: this is the first time I've seen continue to a label in real Dart code). My intuition is that there should be a simpler way to express this.

Would this idea work?

Run through the list of lines looking for reflink definitions. When you find one, extract it and replace those elements in the list with `null`. Then, at the end, filter out the null elements. That way you don't keep creating and throwing away new lists.

Would that work?